From b10ef8fe182915a8af2a0cc2ebca16d45f4aaba1 Mon Sep 17 00:00:00 2001
From: Murat Sezgin <msezgin@codeaurora.org>
Date: Fri, 2 Jun 2017 10:22:27 -0700
Subject: [PATCH 235/281] bridge: Add flag to force netfilter hook evaluation

Add a sysctl (similar to the ones for iptables/ip6tables) to allow user
space to force evaluation of Netfilter hooks registered with the bridge.
This allows custom module's hooks to be invoked without the
iptables/ip6tables/arptables special processing being enabled.
Without this, any such modules will never have their hooks invoked
unless one of the other flags is set via sysctl.

Change-Id: I2875c77525c96c8f69774ab41d5d77e240d88dbb
Signed-off-by: Murat Sezgin <msezgin@codeaurora.org>
---
 net/bridge/br_netfilter_hooks.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/net/bridge/br_netfilter_hooks.c b/net/bridge/br_netfilter_hooks.c
index 5fea9a75f804..3beb03a50ac8 100644
--- a/net/bridge/br_netfilter_hooks.c
+++ b/net/bridge/br_netfilter_hooks.c
@@ -57,6 +57,7 @@ struct brnf_net {
 	int call_ip6tables;
 	int call_arptables;
 	int call_ebtables;
+	int call_custom;
 
 	/* default value is 0 */
 	int filter_vlan_tagged;
@@ -78,7 +79,7 @@ bool br_netfilter_run_hooks(struct net *net)
 	struct brnf_net *brnf = net_generic(net, brnf_net_id);
 
 	return brnf->call_iptables | brnf->call_ip6tables | brnf->call_arptables |
-		brnf->call_ebtables;
+		brnf->call_ebtables | brnf->call_custom;
 }
 
 static inline __be16 vlan_proto(const struct sk_buff *skb)
@@ -1110,6 +1111,12 @@ static struct ctl_table brnf_table[] = {
 		.mode		= 0644,
 		.proc_handler	= brnf_sysctl_call_tables,
 	},
+	{
+		.procname	= "bridge-nf-call-custom",
+		.maxlen		= sizeof(int),
+		.mode		= 0644,
+		.proc_handler	= brnf_sysctl_call_tables,
+	},
 	{ }
 };
 
@@ -1119,6 +1126,7 @@ static inline void br_netfilter_sysctl_default(struct brnf_net *brnf)
 	brnf->call_ip6tables = 1;
 	brnf->call_arptables = 1;
 	brnf->call_ebtables = 1;
+	brnf->call_custom = 1;
 	brnf->filter_vlan_tagged = 0;
 	brnf->filter_pppoe_tagged = 0;
 	brnf->pass_vlan_indev = 0;
@@ -1143,6 +1151,7 @@ static int br_netfilter_sysctl_init_net(struct net *net)
 	table[4].data = &brnet->filter_pppoe_tagged;
 	table[5].data = &brnet->pass_vlan_indev;
 	table[6].data = &brnet->call_ebtables;
+	table[7].data = &brnet->call_custom;
 
 	br_netfilter_sysctl_default(brnet);
 
-- 
2.17.1

